packages/@aws-cdk/custom-resource-handlers/lib/custom-resources-framework/README.md

```diff
@@ -0,0 +1,87 @@
+# AWS CDK Custom Resource Framework
+
+The CDK custom resource framework is an internal framework developed to establish best practices, runtime enforcement, and consistency for custom resource providers and their associated source code. 
+
+## Custom Resource Framework Concepts
+
+Custom resource providers can be created in one of three forms - `CdkFunction`, `CdkSingletonFunction`, or `CdkCustomResourceProvider`. These three custom resource provider formats are a code generated wrapper around the lambda `Function`, lambda `SingletonFunction`, and `CustomResourceProvider` constructs, respectively. These new CDK prefixed constructs will offer the same behavior and initialization options as their parent minus the ability to configure the following properties during construction:
```

Comment:
This framing around `CdkFunction`, `CdkSingletonFunction` etc is confusing me a little.

As far as I can tell, those exact classes are gone (which is great!) so now these are just **concepts**. That's fine, but as a concept a `CdkFunction` doesn't really mean anything to me.

Can we not give these names after what you would use them for:

* `CdkFunction` -> `BuiltinLambdaFunction` (because I think these are never used as custom resource providers?)
* `SingletonFunction` -> actually confusing to me. Are these usually/always used for custom resource providers? Or no? Are they also built-in Lambda functions that are not actually used as custom resource providers? If true, the naming of this section is confusing, because you're heavily stating it's all about custom resources everywhere, but it might not actually be? Maybe it's just non-Custom Resource Lambdas we provision for you?
* `CdkCustomResourceProvider` -> this should make a distinction between the `core.CustomResourceProvider` and the `custom-resources.CustomResourceProvider`. Maybe name them something like `CoreCustomResource` and `LongStabilizationCustomResource`?

Maybe the flavors of Lambdas we generate here are:

* `RuntimeLambda({ singleton: boolean })`
* `CoreCustomResource`
* `LongStabilizationCustomResource({ onEvent, isComplete })`

Reply:
(I'm a little worried we're only adding more layers of indirection, and it looks like we're only adding complexity instead of taking it away...)

Reply:
(If `CdkFunction`, `CdkSingletonFunction` etc are still in from the previous commit, then I think we can take them out again, and we can just codegen their implementations)